snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -257,3 +278,167 @@ def terminate(self) -> None:
+    """"
+    This strategy is responsible for producing a message for all of the subscriptions
+    scheduled for a given tick.
+
+    The subscriptions to be scheduled are those that:
+    - Have the same partition index as the tick received, and
+    - Are scheduled to be run within the time interval defined by the tick
+    (taking the jitter into account)
+
+    The subscriptions to be scheduled are encoded and produced to the
+    scheduled topic to be picked up by the subscription executor which
+    will run the actual query later.
+
+    When a tick is submitted to this strategy, we produce all of it's
+    subscriptions and place it in a queue without waiting to see if they
+    were succesfully produced or not.
+
+    On each call to poll(), we remove completed futures from the queue.
+    If all scheduled subscription messages for that tick were succesfully
+    produced and the tick indicates it should be committed, we commit offsets.
+    """
+
+    def __init__(
+        self,
+        entity_key: EntityKey,
+        schedulers: Mapping[int, Scheduler[Subscription]],
+        producer: Producer[KafkaPayload],
+        scheduled_topic_spec: KafkaTopicSpec,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__schedulers = schedulers
+        self.__encoder = SubscriptionScheduledTaskEncoder(entity_key)
+        self.__producer = producer
+        self.__scheduled_topic = Topic(scheduled_topic_spec.topic_name)
+        self.__commit = commit
+        self.__closed = False
+
+        # Stores each tick with it's futures
+        self.__queue = ScheduledSubscriptionQueue()
+
+        # Not a hard max
+        self.__max_buffer_size = 10000
+
+    def poll(self) -> None:
+        # Remove completed tasks from the queue and raise if an exception occurred.
+        # This method does not attempt to recover from any exception.
+        # Also commits any offsets required.
+        while True:
+            tick_subscription = self.__queue.peek()
+
+            if tick_subscription is None:
+                break
+
+            if not tick_subscription.subscription_future.done():
+                break
+
+            tick_subscription.subscription_future.result()
+
+            self.__queue.popleft()
+
+            if tick_subscription.should_commit:
+                self.__commit(
+                    {
+                        tick_subscription.tick_message.partition: Position(
+                            tick_subscription.tick_message.offset,
+                            tick_subscription.tick_message.timestamp,
+                        )
+                    }
+                )
+
+    def submit(self, message: Message[CommittableTick]) -> None:
+        assert not self.__closed
+
+        # If queue is full, raise MessageRejected to tell the stream
+        # processor to pause consuming
+        if len(self.__queue) >= self.__max_buffer_size:
+            raise MessageRejected
```

Comment:
nice. 
Please let's ensure we add some metrics about this in arroyo as we are barely using this exception today and we are not monitoring how long we spend sending back pressure